torch/csrc/autograd/python_ir.cpp

```diff
@@ -0,0 +1,126 @@
+#include "torch/csrc/autograd/python_ir.h"
+#include "torch/csrc/utils/python_strings.h"
+
+using namespace torch::autograd;
+
+PyObject* THPNodeClass = nullptr;
+
+PyObject* THPNode_Wrap(const std::shared_ptr<Node>& node)
+{
+  if (!node) {
+    Py_RETURN_NONE;
+  } else {
+    auto type = (PyTypeObject*) THPNodeClass;
+    THPNode* obj = (THPNode*)type->tp_alloc(type, 0);
+    if (obj) {
+      obj->cdata = node;
+    }
+    return (PyObject*) obj;
+  }
+}
+
+static int THPNode_traverse(THPNode *self, visitproc visit, void *arg)
+{
+  if (self->cdata) {
+    if (auto fn = dynamic_cast<PyNode*>(self->cdata.get())) {
+      Py_VISIT(fn->pyobj);
+    }
+  }
+  return 0;
+}
+
+static int THPNode_clear(THPNode *self)
+{
+  if (self->cdata) {
+    if (auto fn = dynamic_cast<PyNode*>(self->cdata.get())) {
+      fn->pyobj = nullptr;
+    }
+  }
+  self->cdata.reset();
+  return 0;
+}
+
+static void THPNode_dealloc(THPNode* self)
+{
+  PyObject_GC_UnTrack(self);
+  THPNode_clear(self);
```

Comment:
It seems a bit weird to me that you're calling `clear` in here. It means that you can create a trace (purely in C), but if you then traverse it in Python for some reason, once the objects will start getting freed, it will also delete all the state in the trace (`fn->pyobj = nullptr`), even though the C objects will still be alive? Generally, you shouldn't manage other objects internal fields. It's cleaner to just release ownership, and let that (maybe) destruct them.

In general even if there's a ref cycle, resetting `self->cdata` will break it.